app/core/chat/task_states/task_state.py

```diff
@@ -0,0 +1,59 @@
+from abc import abstractmethod
+from typing import List, Dict, Union
+
+from app.core.chat.dialogue_structs.intent import IntentEnum
+from app.core.chat.dialogue_structs.slot_mapping import SlotMapping
+from app.core.chat.nlg.nlg import NLG
+
+
+class TaskState:
+    def __init__(self, intent: IntentEnum, slots: List[SlotMapping]):
+        self.intent: IntentEnum = intent
+        self.slots: List[SlotMapping] = slots
+
+    def reset(self):
+        for slot in self.slots:
+            slot.set_certain_values(None)
+
+    def update_certain_slots(self, slots_values: Dict[str, Union[str, List[str]]]):
+        for slot in self.slots:
+            if slot.name not in slots_values:
+                continue
+            slot.set_certain_values(slots_values[slot.name])
+
+    def update_guessed_slots(self, slots_values: Dict[str, Union[str, List[str]]]):
+        for slot in self.slots:
+            if slot.name not in slots_values:
+                continue
+            slot.set_guessed_values(slots_values[slot.name])
+
+    def get_slot_names_with_descriptions(self) -> List[str]:
+        return [f"{slot.name} ({slot.description})" for slot in self.slots]
+
+    @abstractmethod
+    def generate_next_response(self, nlg: NLG) -> str:
+        pass
+
+    @abstractmethod
+    def generate_suggestions(self, nlg: NLG) -> str:
+        pass
+
+    def _get_next_empty_required_slot(self) -> SlotMapping:
```

Comment:
I would return `Optional[SlotMapping]` here and do `if is not None: ... else: ...`